### PR TITLE
Javalab: set up readonly mode

### PR DESCRIPTION
**Pull request by bencodeorg**
[note: most of the changes here are a result of having to move things around in javalabEditorTest]

Javalab goes into a "read only" mode when teachers look at their student's work. This same mode will be needed for students to review each other's work. This PR:

- Disables changing the code in the editor when in read only mode.
- Puts disabled styling (gray, no hover highlight, no clickable cursor, no onclick) on the editor header buttons (Commit Code, Backpack, etc.), the "Finish" button, as well as the "Manage Assets" button that is visible in the visualization column in Theater.

Note that:
- Running code is still allowed
- Read only mode is set on page load and never changed.

Editing mode (default):

image

Read only mode:

image

## Testing story

Tested manually and added tests to cover the new functionality.

### CODE REVIEW

File: apps/package.json [hunk 1]
```diff
@@ -233,7 +233,7 @@
     "@codemirror/matchbrackets": "^0.18.0",
     "@codemirror/rectangular-selection": "^0.18.0",
     "@codemirror/search": "^0.18.0",
-    "@codemirror/state": "^0.18.0",
+    "@codemirror/state": "^0.18.7",
```
Comment (bencodeorg) [inline apps/package.json:236]: 0.18.7 is where a change that prevents copy/pasting into the editor was introduced:

https://github.com/codemirror/view/commit/b06ae0399118f2ad2292d131f524b289e2acb090

File: apps/test/unit/javalab/JavalabEditorTest.js [hunk 2]
```diff
@@ -56,338 +57,419 @@ describe('Java Lab Editor Test', () => {
...
+            .props().isDisabled;
+
+          expect(isButtonDisabled).to.be.false;
+        });
       });
     });
+  });
 
-    it('can delete the only file in the editor', () => {
+  describe('View Only Mode', () => {
```
Comment (bencodeorg) [inline apps/test/unit/javalab/JavalabEditorTest.js:443]: I had to move existing tests into their own `describe` block b/c I wanted to set a page constant for these new tests, which can't be reset. Most of the changes to this file are just from that move -- the tests in the 'View Only Mode' block are new.

File: apps/test/unit/javalab/JavalabEditorTest.js [hunk 3]
```diff
@@ -56,338 +57,417 @@ describe('Java Lab Editor Test', () => {
...
+        expect(javalabEditor.state.fileMetadata).to.deep.equal({});
       });
     });
-  });
 
-  describe('File type updates', () => {
-    it('updates visibility', () => {
-      const first = 'Class0.java';
-      const second = 'Class1.java';
+    it('is editable', () => {
```
Comment (bencodeorg) [inline apps/test/unit/javalab/JavalabEditorTest.js:421]: the 'is editable' and 'header buttons are enabled' tests are new tests as well.

File: apps/src/javalab/JavalabButton.jsx [hunk 4]
```diff
@@ -56,6 +66,12 @@ const styles = {
   horizontal: {
     flexDirection: 'row'
   },
+  disabled: {
+    backgroundColor: color.light_gray,
+    borderColor: color.light_gray,
+    ':hover': null,
+    cursor: 'default'
+  },
```
Comment (sanchitmalhotra126) [inline apps/src/javalab/JavalabButton.jsx:74]: Question: do we have a global "disabled" style somewhere that other apps also use, or is this specific to Javalab?

File: apps/i18n/javalab/en_us.json [hunk 5]
```diff
@@ -21,6 +21,7 @@
   "javaExtensionMissing": "Invalid file name. File names must end in '.java'.",
   "newFile": "New File",
   "noMainMethod": "Error: your program does not contain a main method.",
+  "readonlyWorkspaceHeader": "View only",
```
Comment (bencodeorg) [inline apps/i18n/javalab/en_us.json]: I copied this from common/en_us, is that bad practice? I felt like importing all of the strings in common/en_us seems like overkill to get this one string, is there a way to not do that?

File: apps/src/javalab/JavalabButton.jsx (hunk 4 above)
Comment (bencodeorg) [reply apps/src/javalab/JavalabButton.jsx:74]: mmm good question...I have not as much experience in our front end so I may not the best person to answer this, but I think the short answer is that it varies... 

We have this `Button` component which has [disabled styling built in](https://github.com/code-dot-org/code-dot-org/blob/staging/apps/src/templates/Button.jsx#L281), but we are not using it in JavalabButton at the moment (listed as a to do elsewhere in this file).

It looks like we do have some [standard disabled styling](https://github.com/code-dot-org/code-dot-org/blob/staging/dashboard/app/assets/stylesheets/application.scss#L3036) here for (HTML) buttons, which would probably work in this case?

The problem is that 3 of the 6 buttons I was trying to style are not `JavalabButton` but `PaneButton`, which are rendered as divs and I could not figure out how to apply disabled styling without my own custom styles.

File: apps/i18n/javalab/en_us.json (hunk 5 above)
Comment (molly-moen) [reply apps/i18n/javalab/en_us.json]: I'm not sure, but it seems reasonable we will include other common strings in the future, so if other labs are importing them it should be ok to do it here too.

**Review by molly-moen — APPROVED** [review]
Looks good!

File: apps/src/javalab/JavalabButton.jsx (hunk 4 above)
Comment (sanchitmalhotra126) [reply apps/src/javalab/JavalabButton.jsx:74]: I see yeah, might be a big effort to consolidate all of these for now

File: apps/i18n/javalab/en_us.json (hunk 5 above)
Comment (jmkulwik) [reply apps/i18n/javalab/en_us.json]: +1 to Molly's comment. We should generally see if the string exists in common first and use that where possible. If we do it that way, the string will already be translated, and it'll be less work for our translators. The only time I'd probably make a new key is if the string already exists in common the key is very different than our use case. i.e. If the key was something like `"studentGradingCommentPlaceholder": "View only"` we'd probably want to make our own.

File: apps/package.json (hunk 1 above)
Comment (jmkulwik) [reply apps/package.json:236]: 🎉

File: apps/src/javalab/JavalabEditor.jsx [hunk 6]
```diff
@@ -504,9 +515,14 @@ class JavalabEditor extends React.Component {
             isRtl={false}
             label={javalabMsg.newFile()}
             leftJustified
+            isDisabled={isReadOnlyWorkspace}
           />
           <PaneSection style={styles.backpackSection}>
-            <Backpack isDarkMode={isDarkMode} />
+            <Backpack
+              id={'javalab-editor-backpack'}
```
Comment (jmkulwik) [inline apps/src/javalab/JavalabEditor.jsx:522]: What does this ID get used for?

Comment (bencodeorg) [reply apps/src/javalab/JavalabEditor.jsx:522]: Just in a unit test -- the other three buttons in the header have IDs, and I thought it made more sense to be able to loop through IDs than have 3 with IDs and one with some other Enzyme selector:

https://github.com/code-dot-org/code-dot-org/blob/a4014906ac25d5e897dbb309993b38ea8d2406e4/apps/test/unit/javalab/JavalabEditorTest.js#L60

**Pull request merged**